vscode-lean4/src/leanclient.ts

```diff
@@ -94,33 +100,96 @@ export class LeanClient implements Disposable {
     }
 
+    getWorkspaceUri() : Uri {
+        let documentUri : Uri = null;
+        let rootPath : Uri = null;
+        if (window.activeTextEditor)
+        {
+            documentUri = window.activeTextEditor.document.uri
+            const folder = workspace.getWorkspaceFolder(documentUri)
+            if (folder) {
+                rootPath = folder.uri;
+            }
+        }
+        if (!rootPath) {
+            const workspaceFolders = workspace.workspaceFolders;
+            if (workspaceFolders && workspaceFolders.length > 0) {
+                rootPath = workspaceFolders[0].uri;
+            }
+        }
+        if (!rootPath) {
+            return Uri.parse('file:///' + cwd());
+        }
+
+        return rootPath;
+    }
+
     async restart(): Promise<void> {
         this.restartingEmitter.fire(undefined)
 
         if (this.isStarted()) {
             await this.stop()
         }
 
-        this.executable = this.storageManager.getLeanPath();
-        if (!this.executable) this.executable = executablePath();
+        this.toolchainPath = this.storageManager.getLeanPath();
+        if (!this.toolchainPath) this.toolchainPath = toolchainPath();
+
         const version = this.storageManager.getLeanVersion();
         const env = addServerEnvPaths(process.env);
         if (serverLoggingEnabled()) {
             env.LEAN_SERVER_LOG_DIR = serverLoggingPath()
         }
 
-        let options = ['--server']
-        if (version) {
-            // user is requesting an explicit version for this workspace.
-            options = ['+' + version, '--server']
+        let cmd = (this.toolchainPath) ? join(this.toolchainPath, 'lake') : 'lake';
+
+        const folder = this.getWorkspaceUri();
+        // check if the lake process will start.
+        let useLake = true;
+        const lakefile = Uri.joinPath(folder, 'lakefile.lean').toString()
+        if (!fs.existsSync(new URL(lakefile))) {
+            useLake = false;
         }
-        if (workspace.workspaceFolders && workspace.workspaceFolders[0]) {
-            // Add workspace folder name to command-line so that it shows up in `ps aux`.
-            options.push('' + workspace.workspaceFolders[0].uri.toString())
+
+        // This is a faster way of finding out lake doesn't work in the
+        // current workspace.  The LanguageClient is much slower because it does
+        // 5 retries and everything... but this testExecute implementation is
+        // not perfect either... would be nice to have a "canExecute" command line
+        // option or something...
+        if (useLake) {
+            // First check we have a version of lake that supports "lake serve"
+            const lakeVersion = await batchExecute(cmd, ['--version'], folder.fsPath, null);
+            const v4 = new Version('3.0.0')
+            const actual = new Version(this.removePrefix(lakeVersion, 'Lake version '))
+            if (actual.compare(v4) >= 0) {
```

Comment:
No need to reinvent the wheel here.

```suggestion
            const actual = this.removePrefix(lakeVersion, 'Lake version ')
            if (semver.gte(actual, '3.0.0')) {
```